### PR TITLE
[Prototype] Illustrate what generated polymorphic models should look like

### PR DESCRIPTION
**Pull request by annelo-msft**
An initial design proposal for polymorphic models was made in https://github.com/Azure/autorest.csharp/issues/2361, and discussed in the 08/11/2022 C# DPG design meeting.  This PR illustrates the result after making updates suggested in that meeting.  The specific commit containing the updates is here: https://github.com/Azure/azure-sdk-for-net/pull/30718/commits/36dcf11b1c067fcd4cc75fc3011ceaeb28a51e5e

### CODE REVIEW

File: sdk/template/Azure.Template/src/Generated/Models/DerivedFromBaseClassWithDiscriminatorA.Serialization.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace Azure.Template.Models
+{
+    public partial class DerivedFromBaseClassWithDiscriminatorA
+    {
+        internal override void WriteCore(Utf8JsonWriter writer)
+        {
+            writer.WriteStartObject();
+            writer.WritePropertyName("discriminatorProperty");
+            writer.WriteStringValue(DiscriminatorProperty);
```
Comment (AlexanderSher) [inline sdk/template/Azure.Template/src/Generated/Models/DerivedFromBaseClassWithDiscriminatorA.Serialization.cs:18]: I still don't understand why do we need `DiscriminatorProperty`. What prevents us from hardcoding it:
```cs
writer.WritePropertyName("discriminatorProperty");
writer.WriteStringValue("A");
```

File: sdk/template/Azure.Template/src/Generated/Models/BaseClassWithDiscriminator.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        /// <param name="discriminatorProperty"> Discriminator property for BaseClassWithDiscriminator. </param>
+        /// <exception cref="ArgumentNullException"> <paramref name="baseClassProperty"/> is null. </exception>
+        // Note: internal constructor to prevent end-user derived types
+        // Note: When we add discriminator to internal initialization constructor, it becomes the serialization constructor.
+        internal BaseClassWithDiscriminator(string baseClassProperty, string discriminatorProperty) : base(baseClassProperty)
+        {
+            Argument.AssertNotNull(baseClassProperty, nameof(baseClassProperty));
+            Argument.AssertNotNullOrEmpty(discriminatorProperty, nameof(discriminatorProperty));
+
+            if (baseClassProperty == null)
```
Comment (AlexanderSher) [inline sdk/template/Azure.Template/src/Generated/Models/BaseClassWithDiscriminator.cs:28]: `baseClassProperty` has been checked for null already in `Argument.AssertNotNull`

File: sdk/template/Azure.Template/src/Generated/Models/BaseClassWithDiscriminator.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+#nullable disable
+
+using System;
+using Azure.Core;
+
+namespace Azure.Template.Models
+{
+    /// <summary> Example base class that has a discriminator property. </summary>
+    // Note: base classes are abstract for polymorphic types
+    public abstract partial class BaseClassWithDiscriminator : BaseClass
```
Comment (AlexanderSher) [inline sdk/template/Azure.Template/src/Generated/Models/BaseClassWithDiscriminator.cs:15]: @KrzysztofCwalina @tg-msft @lmazuel Can you confirm that base class that has discriminator property should never be instantiated?

File: sdk/template/Azure.Template/src/Generated/Models/BaseClassWithDiscriminator.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        /// <summary> Initializes a new instance of BaseClassWithDiscriminator. </summary>
+        /// <param name="baseClassProperty"> An example property. </param>
+        /// <param name="discriminatorProperty"> Discriminator property for BaseClassWithDiscriminator. </param>
+        /// <exception cref="ArgumentNullException"> <paramref name="baseClassProperty"/> is null. </exception>
+        // Note: internal constructor to prevent end-user derived types
+        // Note: When we add discriminator to internal initialization constructor, it becomes the serialization constructor.
+        internal BaseClassWithDiscriminator(string baseClassProperty, string discriminatorProperty) : base(baseClassProperty)
```
Comment (AlexanderSher) [inline sdk/template/Azure.Template/src/Generated/Models/BaseClassWithDiscriminator.cs:23]: should it be `protected` or `private protected`?

File: sdk/template/Azure.Template/src/Generated/Models/DerivedFromBaseClassWithDiscriminatorA.Serialization.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+#nullable disable
+
+using System.Text.Json;
+
+namespace Azure.Template.Models
+{
+    public partial class DerivedFromBaseClassWithDiscriminatorA
+    {
+        internal override void WriteCore(Utf8JsonWriter writer)
```
Comment (AlexanderSher) [inline sdk/template/Azure.Template/src/Generated/Models/DerivedFromBaseClassWithDiscriminatorA.Serialization.cs:14]: Should this be marked `sealed`?

File: sdk/template/Azure.Template/src/Generated/Models/DerivedFromBaseClassWithDiscriminatorUnknown.Serialization.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            writer.WritePropertyName("baseClassProperty");
+            writer.WriteStringValue(BaseClassProperty);
+            writer.WriteEndObject();
+        }
+
+        internal static DerivedFromBaseClassWithUnknownDiscriminator DeserializeDerivedFromBaseClassWithDiscriminatorUnknown(JsonElement element)
+        {
+            string discriminatorProperty = default;
+            string baseClassProperty = default;
+            foreach (var property in element.EnumerateObject())
```
Comment (AlexanderSher) [inline sdk/template/Azure.Template/src/Generated/Models/DerivedFromBaseClassWithDiscriminatorUnknown.Serialization.cs:28]: If we want to use this instance in roundtrip scenario, we must preserve all the properties that are available for the element.

File: sdk/template/Azure.Template/api/Azure.Template.netstandard2.0.cs [hunk 7]
```diff
@@ -1,35 +1,39 @@
...
-        public virtual Azure.Core.Pipeline.HttpPipeline Pipeline { get { throw null; } }
-        public virtual Azure.Response GetSecret(string secretName, Azure.RequestContext context = null) { throw null; }
-        public virtual System.Threading.Tasks.Task<Azure.Response> GetSecretAsync(string secretName, Azure.RequestContext context = null) { throw null; }
-        public virtual Azure.Response<Azure.Template.Models.SecretBundle> GetSecretValue(string secretName, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
-        public virtual System.Threading.Tasks.Task<Azure.Response<Azure.Template.Models.SecretBundle>> GetSecretValueAsync(string secretName, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        internal BaseClass() { }
+        public string BaseClassProperty { get { throw null; } set { } }
     }
-    public partial class TemplateClientOptions : Azure.Core.ClientOptions
+    public abstract partial class BaseClassWithDiscriminator : Azure.Template.Models.BaseClass
```
Comment (KrzysztofCwalina) [inline sdk/template/Azure.Template/api/Azure.Template.netstandard2.0.cs:8]: Why does this class not actually have the discriminator property?

File: sdk/template/Azure.Template/src/Generated/Models/BaseClassWithDiscriminator.cs (hunk 2 above)
Comment (annelo-msft) [reply sdk/template/Azure.Template/src/Generated/Models/BaseClassWithDiscriminator.cs:28]: Copy/paste/delete miss!

File: sdk/template/Azure.Template/src/Generated/Models/BaseClassWithDiscriminator.cs (hunk 3 above)
Comment (annelo-msft) [reply sdk/template/Azure.Template/src/Generated/Models/BaseClassWithDiscriminator.cs:15]: Management plan is doing this.  The client can still return an instance of the base class with an unknown discriminator using the "internal unknown" type, which in this PR is `DerivedFromBaseClassWithUnknownDiscriminator`

File: sdk/template/Azure.Template/api/Azure.Template.netstandard2.0.cs (hunk 7 above)
Comment (annelo-msft) [reply sdk/template/Azure.Template/api/Azure.Template.netstandard2.0.cs:8]: It does - it's just internal, so you don't see it in the public API.

File: sdk/template/Azure.Template/src/Generated/Models/BaseClassWithDiscriminator.cs (hunk 3 above)
Comment (lmazuel) [reply sdk/template/Azure.Template/src/Generated/Models/BaseClassWithDiscriminator.cs:15]: Yes, it happens that you need to return the base class and you don't know the discriminator (sorry for the late answer)

Comment (annelo-msft) [reply sdk/template/Azure.Template/src/Generated/Models/BaseClassWithDiscriminator.cs:15]: @lmazuel - Thanks!  Can you help us understand the scenario you're describing?

**Pull request closed without merging**